robotframework-ls/src/robotframework_ls/impl/flow_explorer_model_builder.py

```diff
@@ -89,64 +142,92 @@ def _build_hierarchy(
-                    if keyword_found.library_name:
-                        keyword[
-                            "name"
-                        ] = f"{keyword_usage_node.keyword} ({keyword_found.library_name})"
-                    elif keyword_found.resource_name:
-                        keyword[
-                            "name"
-                        ] = f"{keyword_usage_node.keyword} ({keyword_found.resource_name})"
-
-                    # If it was found in a library we don't recurse anymore.
-                    keyword_ast = keyword_found.keyword_ast
-                    if keyword_ast is None:
-                        continue
-                    definition_completion_context = keyword_found.completion_context
-                    if definition_completion_context is None:
-                        continue
-
-                    suite_name = _compute_suite_name(definition_completion_context)
-                    # Ok, it isn't a library keyword (as we have its AST). Keep recursing.
-                    for node_info in ast_utils.iter_all_nodes(
-                        keyword_ast, recursive=False
-                    ):
-                        _build_hierarchy(
-                            definition_completion_context,
-                            node_info.stack,
-                            node_info.node,
-                            suite_name,
-                            keyword_body,
-                            memo,
-                        )
+                    # Fallback name if we don't know where it's defined.
+                    keyword["name"] = f"{keyword_usage_node.keyword}"
 
+                    if definitions:
+                        # Use the first one
+                        definition = next(iter(definitions))
+                        keyword_found: IKeywordFound = definition.keyword_found
+                        if keyword_found.library_name:
+                            keyword[
+                                "name"
+                            ] = f"{keyword_usage_node.keyword} ({keyword_found.library_name.lower()})"
+                        elif keyword_found.resource_name:
+                            keyword[
+                                "name"
+                            ] = f"{keyword_usage_node.keyword} ({keyword_found.resource_name.lower()})"
+
+                        # If it was found in a library we don't recurse anymore.
+                        keyword_ast = keyword_found.keyword_ast
+                        if keyword_ast is None:
+                            continue
+                        definition_completion_context = keyword_found.completion_context
+                        if definition_completion_context is None:
+                            continue
+                        # If found in recursion stack we don't recurse anymore.
+                        if recursion_stack and recursion_stack.check(keyword["name"]):
```

Comment:
I don't think you need to check `if recursion_stack` part here (it can't be None as far as I can see so that'd always evaluate to `True` right?).